server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -155,18 +156,21 @@ protected final boolean countCanUseMetadata(FiltersAggregator.Counter counter, B
      */
     QueryToFilterAdapter<?> union(Query extraQuery) throws IOException {
         /*
+         * Wrap with a ConstantScoreQuery to enables a few more rewrite
+         * rules as of Lucene 9.2.
          * It'd be *wonderful* if Lucene could do fancy optimizations
-         * when merging queries but it doesn't at the moment. Admittedly,
-         * we have a much more limited problem. We don't care about score
-         * here at all. We know which queries its worth spending time to
-         * optimize because we know which aggs rewrite into this one.
+         * when merging queries like combining ranges but it doesn't at
+         * the moment. Admittedly, we have a much more limited problem.
+         * We don't care about score here at all. We know which queries
+         * it's worth spending time to optimize because we know which aggs
+         * rewrite into this one.
          */
-        extraQuery = searcher().rewrite(extraQuery);
-        if (extraQuery instanceof MatchAllDocsQuery) {
+        extraQuery = searcher().rewrite(new ConstantScoreQuery(extraQuery));
+        Query unwrappedExtraQuery = unwrap(extraQuery);
+        if (unwrappedExtraQuery instanceof MatchAllDocsQuery) {
```

Comment:
I'm checking the unwrapped query now instead of the original query, is this correct?

Reply:
:+1: